### PR TITLE
[Clang][CTAD][NFC] Unify transformTemplateParameter()

### PR DESCRIPTION
**Pull request by zyn0217**
We ended up having two transformTemplateParameter() after CTAD for type aliases was landed. This patch cleans them up and allows them to share one implementation.

As a bonus, this also uses getDepthAndIndex() in preference to getTemplateParameter{Depth,Index}().

### CODE REVIEW

**Comment by zyn0217** [comment]
The flaky test `basic-project.test` failure is unrelated -- I didn't rebase my branch onto 88549cf4

File: clang/lib/Sema/SemaTemplateDeductionGuide.cpp [hunk 1]
```diff
@@ -284,6 +283,42 @@ transformTemplateParam(Sema &SemaRef, DeclContext *DC,
...
+
+unsigned getTemplateParameterIndex(NamedDecl *TemplateParam) {
+  if (auto *TTP = dyn_cast<TemplateTypeParmDecl>(TemplateParam))
+    return TTP->getIndex();
+  if (auto *TTP = dyn_cast<TemplateTemplateParmDecl>(TemplateParam))
+    return TTP->getIndex();
+  if (auto *NTTP = dyn_cast<NonTypeTemplateParmDecl>(TemplateParam))
+    return NTTP->getIndex();
+  llvm_unreachable("Unhandled template parameter types");
+}
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaTemplateDeductionGuide.cpp:320]: Given we already have `getDepthAndIndex` - we can simplify that further

**Review by cor3ntin — COMMENTED** [review]
Thanks for cleaning that up

File: clang/lib/Sema/SemaTemplateDeductionGuide.cpp (hunk 1 above)
Comment (zyn0217) [reply clang/lib/Sema/SemaTemplateDeductionGuide.cpp]: Thanks! I should've remembered we have used it in `CheckParameterPacksForExpansion()`!

**Review by cor3ntin — APPROVED** [review]
I think this looks good, nice cleanup.

**Pull request merged**